In [105]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from scipy import stats



1.  A bank found that the average number of cars waiting during the noon hour at a drive-up window follows a Poisson distribution with a mean of 2 cars. Make a chart of this distribution and answer these questions concerning the probability of cars waiting at the drive-up window.

What is the probability that no cars drive up in the noon hour?

In [36]:
λ = 2

cars = stats.poisson(λ)

What is the probability that 3 or more cars come through the drive through?

In [37]:
cars.sf(2)

0.32332358381693654

How likely is it that the drive through gets at least 1 car?

In [5]:
cars.cdf(1)

0.40600584970983794

2.  Grades of State University graduates are normally distributed with a mean of 3.0 and a standard deviation of .3. Calculate the following:

In [39]:
μ = 3.0
σ = .3

gpa = stats.norm(μ, σ)

What grade point average is required to be in the top 5% of the graduating class?

In [40]:
gpa.ppf(.95)

3.4934560880854417

In [41]:
gpa.isf(.05)

3.4934560880854417

What GPA constitutes the bottom 15% of the class?

In [42]:
gpa.ppf(.15)

2.689069983151863

In [43]:
gpa.isf(.85)

2.689069983151863

An eccentric alumnus left scholarship money for students in the third decile from the bottom of their class. Determine the range of the third decile. Would a student with a 2.8 grade point average qualify for this scholarship?


In [44]:
gpa.ppf(.2)

2.7475136299281258

In [45]:
gpa.ppf(.3)

2.842679846187588

The GPA range is 2.748 - 2.843

A 2.8 GPA student would qualify!

If I have a GPA of 3.5, what percentile am I in?

In [46]:
gpa.cdf(3.5)

0.9522096477271853

3.  A marketing website has an average click-through rate of 2%. One day they observe 4326 visitors and 97 click-throughs. How likely is it that this many people or more click through?

In [47]:
click = stats.binom(4326, .02)

In [31]:
click.sf(96)

0.13975823631416445

4.  You are working on some statistics homework consisting of 100 questions where all of the answers are a probability rounded to the hundreths place. Looking to save time, you put down random probabilities as the answer to each question.

What is the probability that at least one of your first 60 answers is correct?

In [34]:
hw_distribution = stats.binom(60, .01)
hw_distribution


In [35]:
hw_distribution.sf(0)

0.45284335760923855

5. The codeup staff tends to get upset when the student break area is not cleaned up. Suppose that there's a 3% chance that any one student cleans the break area when they visit it, and, on any given day, about 90% of the 3 active cohorts of 22 students visit the break area. How likely is it that the break area gets cleaned up each day? How likely is it that it goes two days without getting cleaned up? All week?

In [66]:
clean = stats.binom(60, .03)
# students = stats.poisson()

In [65]:
clean.sf(0)

0.8342199288437352

In [89]:
two_days_clean = stats.binom(119, 0.03)
two_days_clean.pmf(0)

0.026658541032798848

In [86]:
all_week = stats.binom(297, 0.03)
all_week.pmf(0)

0.00011781621791055051

6. You want to get lunch at La Panaderia, but notice that the line is usually very long at lunchtime. After several weeks of careful observation, you notice that the average number of people in line when your lunch break starts is normally distributed with a mean of 15 and standard deviation of 3. If it takes 2 minutes for each person to order, and 10 minutes from ordering to getting your food, what is the likelihood that you have at least 15 minutes left to eat your food before you have to go back to class? Assume you have one hour for lunch, and ignore travel time to and from La Panaderia.

In [100]:
lunch = stats.norm(15, 3)
#Assuming one hour lunch
#2x+10 must be less than 45 in order to have time to eat so x must be less than 17.5
lunch.cdf(16.5)
#we have a 69.146 percent chance we'll be able to finish lunch

0.6914624612740131

In [101]:
minutes = stats.norm(30, 6)
minutes.cdf(33)

0.6914624612740131

7.  Connect to the employees database and find the average salary of current employees, along with the standard deviation. For the following questions, calculate the answer based on modeling the employees salaries with a normal distribution defined by the calculated mean and standard deviation then compare this answer to the actual values present in the salaries dataset.

In [113]:
from env import get_db_url



query = '''
SELECT salary
FROM salaries
WHERE to_date > curdate(); 

'''

salaries = pd.read_sql(query, get_db_url('employees'))

In [115]:
salaries.describe()

salary
count  240124.000000
mean    72012.235857
std     17309.995380
min     38623.000000
25%     59048.750000
50%     69805.000000
75%     82546.000000
max    158220.000000

What percent of employees earn less than 60,000?

In [116]:
sal = stats.norm(72012.235857, 17309.995380)

In [117]:
sal.cdf(60000)

0.2438574248045388

What percent of employees earn more than 95,000?

In [118]:
sal.sf(94999.99)

0.0920882874148749

What percent of employees earn between 65,000 and 80,000?

In [120]:
sal.cdf(80000)-sal.cdf(65000)

0.33506056854908584

In [123]:
np.diff(sal.cdf([65000, 80000]))

array([0.33506057])

What do the top 5% of employees make?

In [125]:
sal.isf(.05)

100484.64454030624

Hint If you're looking at this exercise and wondering "How do I get pandas to talk the database, again?", remember that you'll need 3 things: your .gitignore, your env.py, and to use pd.read_sql. Copy over your .gitignore and env.py from your data science libraries exercises folder, and connect to the employees database like so: